### DISCUSSION TITLE
H5P Defaults

### DISCUSSION THREAD

**Discussion by lukecarr** [Lumieducation/Lumi#1706] [Q&A]
Hi,

Does anyone know if it's possible to configure some default values for H5P activity settings?

I'm referring to settings such as these within a Question Set:

![image](https://user-images.githubusercontent.com/24438483/126157381-a6395d9c-b838-4a1f-931e-61573c1268fd.png)

I'm happy to get my hands dirty if it means forking and making some changes, but any pointers on where to start would be greatly appreciated!

**Comment by sr258** [comment]
This would certainly be a very useful feature and I've already thought about this for some time! I'd be happy about a PR!

To achieve this, you'll have to modify the semantic structure of the library in question, which also includes default values for these objects. You can find the official documentation page on how the semantics are structured here: https://h5p.org/semantics

Our underlying [h5p-nodejs-library](https://github.com/lumieducation/h5p-nodejs-library) library supports changing the semantic structure on-the-fly, but Lumi does not make this functionality available in the GUI yet. There is documentation on how this can be done with h5p-nodejs-library here: https://docs.lumi.education/advanced-usage/customization#changing-the-semantics-of-individual-libraries 

Basically, you'll have to add an implementation of alterLibrarySemantics that is generic and can be configured by the user through some kind of special page. This could be put into another page in the settings dialog.

If you need further guidance, drop me a line!

**Reply by sr258** [reply]
Note: If you are only interested in creating a forked Lumi version that behaves differently for some content types, it's not necessary to add all the GUI functionality. All you'll have to do is add an alterLibrarySemantics hook [here](https://github.com/Lumieducation/Lumi/blob/6d93d5f95e6547997d0332f5d947414a3869062c/server/src/boot/createH5PEditor.ts#L47) that performs the changes to the semantic structure you need.
